viewer-prototype/src/node/trace-server-service.ts

```diff
@@ -4,11 +4,31 @@ import { TraceServerConfigService } from '../common/trace-server-config';
 @injectable()
 export class TraceServerServiceImpl implements TraceServerConfigService {
 
-    async startTraceServer(path: string | undefined, port: number | undefined): Promise<void> {
-        spawn(`${path}`, ['-vmargs', `-Dtraceserver.port=${port}`]);
+    startTraceServer(path: string | undefined, port: number | undefined): Promise<string> {
+        return new Promise<string>((resolve, reject) => {
+            const server = spawn(`${path}`, ['-vmargs', `-Dtraceserver.port=${port}`]);
+            server.on('error', err => {
+                reject(err.message);
+            });
+            server.on('message', () => {
+                resolve('success');
+            });
```

Comment:
I'm surprised this even works. From what I understand, the `child_process` event `message` is fired when the spawned process communicates back using Node's special IPC (which requires an extra pipe to be setup). Are you sure this code works, and if so why?

Reply:
See:
- https://nodejs.org/docs/latest-v12.x/api/child_process.html#child_process_event_message
- https://nodejs.org/docs/latest-v12.x/api/process.html#process_process_send_message_sendhandle_options_callback

Reply:
@marechal-p  Yes, I commented on #233 the same. The resolve method will never be called. @ankusht-work the reason why the status changes is the periodic scheduled ping which notices the whether the server is up or down.

The thing is that when the server starts successfully, then the child process doesn't return. So, the spawn won't return and the promise won't be resolve successfully. Only error cases can be caught if we add server.on('exit') and check the exit code returned. Not sure how we should handle the successful case. How does is it done with the language server, I wonder?

Reply:
So we agree that this snippet is bogus, cool.

I usually check for the PID after the spawn call to check if the process started. This doesn't mean that there won't be an error later, but it's a best effort. From then on you should listen for exit anyway and you'll know if a crash occurred thanks to the return code.

```ts
const child = spawn(...);
child.on('error', reject);
// If the executable is not found, PID will be undefined:
if (typeof child.pid === 'number') {
    resolve();
}
```

Reply:
Else you could wait for a specific output line to show up. I'm always on the fence for this kind of stdout reading but it should be rather reliable depending on what the server does. If you try doing that you should look at [`readline`](https://nodejs.org/docs/latest-v12.x/api/readline.html#readline_readline_createinterface_options)

Reply:
More clarification as to how this is working is provided in [226](https://github.com/theia-ide/theia-trace-extension/pull/233), I have not been able to figure out yet a proper solution to handle the resolve case. 
Below are the assumptions we are running with for now:
1) If the spawn itself fails (for ex: the path to the executable is incorrect)   
       -----error listener will return the error message back to the front-end.
2) If the port is busy:
       -----exit listener will return the exit code back 
3)  If the process generates any output or message, consider that as a success.